### ISSUE TITLE
Websocket Secure support

### ISSUE THREAD

**Issue by BoarK** [lipp/lua-websockets#18]
Hi lipp,

In client*.lua I see the following code snippet:

if protocol ~= 'ws' then
   error('Protocol not supported:'..protocol)
end

There is no support for wss:// ?  Is secure connection support on the roadmap?  A non-secured websocket connection will run into proxy problems.

Nathan.

**Comment by ckilimci** [comment]
Hi @lipp 

I guess luasec ready by https://github.com/brunoos/luasec/commit/84cb83b92f2c695064aa0a72e0eee0c34fd46c15

Do you plan to add secure support for now? If you are or not, thank you to create such an awesome package.

**Comment by lipp** [comment]
@ckilimci thx! 

just out of curiosity: what application / environment are you targeting with lua + websockets?
i guess it is something "embedded" (that's why you choose lua over e.g. node.js).

for wss servers consider this:
to employ secure websockets you need certificates, which are bound to domain names (!!!). in my environment, lua is used on embedded systems with limited ressources. unfortunately these embedded systems are not public (have no domain at all) --> using wss is pointless. you can use wss without true certs/domain pairs but this will either bother the user with warnings or will not work in future browser versions which might have tighter security policies.


for wss clients lua+websockets can make perfect sense.
i will have a look at what is a good starting point to implement wss for clients.

**Comment by ckilimci** [comment]
You are right @lipp I wanted to use lua+websocket for wss client in embedded.

**Comment by suoloordi** [comment]
Hello @lipp 
Do you have any updates about this? An ETA maybe? I was looking for a ws client to use on OpenWrt and I was really glad to have found your work. It would be amazing if wss was supported.

**Comment by dakhota** [comment]
Hi @lipp,

It seems [luasec's](https://github.com/brunoos/luasec) ssl module can wrap tcp connections. I tried to add it to client_ev.lua. I think somethig like this should go somewhere after [`if connected ...`](https://github.com/lipp/lua-websockets/blob/61cca37c9257ce68df6cc94300d5a191a365687d/src/websocket/client_ev.lua#L194):

```lua
if protocol == 'wss' then
  sock:timeout(1) -- it timed out for me without this
  sock = ssl.wrap(sock,{ mode = 'client', protocol = 'sslv23' }) -- tlsv1 segfaulted when i tried it with a plain tcp socket; may need more params, e.g. cafile = '/etc/...', verify = 'peer', options = 'all'
  sock:dohandshake() -- i don't know if it's needed for the ssl wrapper
  sock:timeout(0) -- turn async back on
end
```

It runs past this code fine, but it doesn't seem to work; I end up with a `"closed"` error from [here](https://github.com/lipp/lua-websockets/blob/61cca37c9257ce68df6cc94300d5a191a365687d/src/websocket/client_ev.lua#L60) I think, and then this happens on the `cleanup()`:

```
CALLBACK FAILED: .../torch/install/share/lua/5.1/websocket/client_ev.lua:48: attempt to call method 'shutdown' (a nil value)
stack traceback:
        .../torch/install/share/lua/5.1/websocket/client_ev.lua:48: in function 'cleanup'
        .../torch/install/share/lua/5.1/websocket/client_ev.lua:64: in function 'on_error'
        .../torch/install/share/lua/5.1/websocket/client_ev.lua:82: in function 'on_err'
        .../torch/install/share/lua/5.1/websocket/ev_common.lua:38: in function 'f'
        .../torch/install/share/lua/5.1/websocket/ev_common.lua:11: in function <.../torch/install/share/lua/5.1/websocket/ev_common.lua:9>
        [C]: in function 'loop'
        ./px.lua:95: in main chunk
        [C]: in function 'dofile'
        .../torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00405d70
```

I started using lua only like a week ago, so I'm quite clueless as for what I'm missing...

**Comment by lipp** [comment]
@dakhota i highly appreciate your efforts!
hard to say what's going wrong. looks ok for me so far...maybe the overriding of `sock` goes wrong? is `ssl.wrap` really returning a sock object with tcp.socket compatible interface?

the error shown in the stacktrace definitely says: wanted to call "sock:shutdown()" but this function was not there. calling shutdown could be caused by something went wrong during previous operations or client:close() was called.

anyhow: i once started myself to do this but never finished.... some stuff was working though... maybe you want to have a look: https://github.com/lipp/lua-websockets/tree/secure-wss

**Comment by dakhota** [comment]
You are right, the object that ssl.wrap returns implements only a subset of the tcp.socket interface. I tried to add a wrapper around it, one that calls the ssl method when it exists but falls back to the original socket's method when it doesn't, but it didn't work (ssl.wrap sets the socket's fd to -1, so I had to set it back to whatever it originally was.)

**Comment by Treeston** [comment]
I was looking to implement a COPAS websocket client for the [Slack RTM API](https://api.slack.com/rtm), which uses secure websockets (`wss://`), in Lua.

Sad to see this isn't supported.

**Comment by NovusTheory** [comment]
Disappointed to find this **still** isn't supported **3 years later**

**Comment by NovusTheory** [comment]
@lipp Disappointed to find this **still** isn't supported **3 years later**

**Comment by lipp** [comment]
@NovusTheory @Treeston sorry for disappointing you. PRs welcome!

**Comment by NovusTheory** [comment]
@lipp #91 and everyone else

**Comment by moteus** [comment]
I write backend for lua-WebSocket based on my libuv binding (both sync and async). And it support wss [wss client example](https://github.com/moteus/lua-lluv-websocket/blob/master/examples/wss_echo.lua).

**Comment by lipp** [comment]
merged @NovusTheory PR with SSL support for sync clients. Thanks!

**Comment by dschu012** [comment]
@lipp the rockspec checks out v2.2. can we get an updated tag/release with #91 ?

**Comment by venkateshveeroju** [comment]
@lipp @dschu012 @moteus @ckilimci Do we have any implementation on this request?
